recipe_scrapers/coopse.py

```diff
@@ -0,0 +1,110 @@
+        recipe_id = None
+        for script in scripts:
+            matches = SCRIPT_PATTERN.search(str(script.string))
+            if matches:
+                data = matches.groupdict()
+                recipe_id = data["recipeId"]
+        data_url = f"https://proxy.api.coop.se/external/recipe/recipes/{recipe_id}?api-version=v1"
+        self.page_data = requests.get(
+            data_url,
+            proxies=proxies,
+            timeout=timeout,
+        ).content
+        self.data = json.loads(self.page_data)
+
+    @classmethod
+    def host(cls):
+        return "coop.se"
+
+    def author(self):
+        return self.data["recipeCreator"]
+
+    def title(self):
+        return self.data["name"]
+
+    def category(self):
+        for c in self.data["categories"]:
+            if len(c["categories"]) > 0:
+                if c["categories"][0] == "Typ av rätt":
+                    return c["categories"][1]
+        return None
+
+    def total_time(self):
+        return int(self.data["totalTime"].replace(" minuter", ""))
+
+    def yields(self):
+        return self.data["numberOfServings"]
+
+    def image(self):
+        return self.data["imageUrl"]
+
+    def ingredients(self):
+        groups = self.ingredient_groups()
+        ingredients = []
+        for g in groups:
+            ingredients.extend(g.ingredients)
+        return ingredients
+
+    def ingredient_groups(self):
+        groups = []
+        for p in self.data["recipePart"]:
+            ingredients = []
+            for i in p["ingredients"]:
+                quantity = i["quantity"]
+                if quantity.endswith(".0"):
+                    quantity = quantity[:-2]
+                unit = i["unit"]
+                if not unit == "":
+                    unit = " {}".format(unit)
+                pre_preparation = i["prePreparation"]
+                if not pre_preparation == "":
+                    pre_preparation = " {}".format(pre_preparation)
+                name = i["name"]
+                if not quantity == "1" and not pre_preparation == " riven":
+                    name = i["ingredient"]["pluralName"]
+                ingredients.append(
+                    "{}{}{} {}{}".format(
+                        quantity, unit, pre_preparation, name, i["postPreparation"]
+                    )
+                )
+
+            groups.append(IngredientGroup(purpose=p["name"], ingredients=ingredients))
+        return groups
+
+    def instructions(self):
+        return "\n".join(
+            list(
+                filter(
+                    lambda x: len(x) > 0,
+                    re.sub(
+                        r"(<p>|<h3>|</h3>)", "", self.data["cookingInstructions"]
+                    ).split("</p>"),
+                )
+            )
+        )
+
+    def ratings(self):
+        return self.data["avgRating"]
+
+    def cuisine(self):
+        return None
+
+    def description(self):
+        return self.data["preamble"]
```

Comment:
Nitpick: because the `preamble` field can sometimes be empty, and because this library generally indicates an absence-of-value by using the Python `None` value instead of the empty string, I think we should perform a small additional check here:

```suggestion
        return self.data["preamble"] or None
```

What this means is: return the `data['preamble']` value if it is non-empty (truthy), but otherwise return the singleton value `None` so that the caller knows that this recipe does not have a description.

Reply:
That's much clearer